src/compiler/instrumented_compiler.rs

```diff
@@ -72,51 +50,18 @@ impl InstrumentedCompiler {
             InstrumentedCompiler::Running {
                 file_name, code, ..
             } => (file_name.as_str(), code.as_str()),
-
-            _ => {
+            InstrumentedCompiler::Finished(_) => {
                 panic!("`file_name_and_code` called on a non-running compiler")
             }
         }
     }
-}
-
-impl Compiler for InstrumentedCompiler {
-    type Output = ChangeSet;
 
-    fn prepare(&mut self) -> Vec<String> {
+    fn settings(&self) -> &GlueCompilerInvocationSettings {
         match self {
-            InstrumentedCompiler::Ready {
-                args,
-                file_name,
-                code,
-            } => {
-                // We're moving data out of &mut self, but this is fine since
-                // we will reassign to self next.
-
-                let args = mem::take(args);
-                let file_name = mem::take(file_name);
-                let code = mem::take(code);
-
-                *self = InstrumentedCompiler::Running { file_name, code };
-                args
-            }
-
-            _ => panic!("`start` called on a non-ready compiler"),
+            InstrumentedCompiler::Running { settings, .. } => settings,
+            _ => panic!("`settings` called on a non-finished compiler"),
         }
     }
-
-    fn run(mut self) -> AnyResult<ChangeSet> {
-        // This call to `to_vec` occurs for borrowing reasons. See the comment
-        // in `StandardCompiler::run` for more.
-
-        let args = self.prepare();
-
-        RunCompiler::new(args.as_slice(), &mut self)
-            .run()
-            .map_err(|_| anyhow!("Failed to compile crate"))?;
-
-        self.finalize()
-    }
```

Comment:
Is there a reason why we don't want this trait implemented for InstrumentedCompiler? It allowed us to cleanly handle the compiler state machine and unifies the behaviour of all our compilers :(